Adds support to read PEM keys

ignasi35 opened the pull request:
References #29036

- [x] basic PEM reading support
- [ ] ~add PEMreading support on akka remote's `SSLEngineProvider`~

ignasi35 commented on build.sbt:
```diff
@@ -315,6 +316,10 @@ lazy val protobufV3 = akkaModule("akka-protobuf-v3")
     test in assembly := {}, // assembly runs tests for unknown reason which introduces another cyclic dependency to packageBin via exportedJars
     description := "Akka Protobuf V3 is a shaded version of the protobuf runtime. Original POM: https://github.com/protocolbuffers/protobuf/blob/v3.9.0/java/pom.xml")
 
+lazy val pemReader =
```
I'm not too happy with this name and, instead, I think this could be `akka-pki`.

I'm avoiding anything with `TLS` or `SSL` in the name to keep this focused and avoid a name that's so generic that this ends up being used to store _anything_.

NOTE: the reason this package exists is to `akka-remote` doesn't have an extra dependency.

ignasi35 replied:
My idea is to make this depend on `akka-remote` so it can provide an implementation of `SSLEngineProvider` that overrides the key and trust loading and uses the pem reader.

patriknw replied:
Is this only intended to be used with akka-remote and not as a general utility from Akka HTTP or Akka gRPC also?

In that case I think we can just include it in akka-remote instead, with the external dependency marked as optional/provided. It can still be a SSLEngineProvider subclass there.

ignasi35 replied:
The PEM reader was brought here for the cloudstate specific use case but being able to read PKCS#1 and PKCS#8 will be very valuable in any TLS use case and sample (e.g. https://github.com/akka/akka-grpc/blob/e460469331e7f2a8b8c58aa6a44cf3eac1cdb0eb/interop-tests/src/main/scala/akka/grpc/interop/AkkaGrpcServerScala.scala#L55-L84).

patriknw replied:
ok, so keep the reader utility in new module, but then it shouldn't depend on akka-remote. The SSLEngineProvider can be in akka-remote with the provided dependencies.

patriknw replied:
Regarding naming, I agree that pem-reader sounds very narrow. Can we broaden it somewhat so it could contain some eventual other utilities in the future? akka-pki might be good?

ignasi35 commented on akka-pem-reader/src/test/scala/akka/pki/pem/PEMPrivateKeyLoaderSpec.scala:
```diff
@@ -0,0 +1,38 @@
+
+  }
+
+}
```
All the code in `pem-reader/src` is brought in _verbatim_ (except for package names and copyright headers).

ignasi35 commented on akka-pem-reader/src/main/scala/akka/remote/artery/tcp/PEMConfigSSLEngineProvider.scala:
```diff
@@ -0,0 +1,99 @@
+import javax.net.ssl.TrustManager
+import javax.net.ssl.TrustManagerFactory
+
+class PEMConfigSSLEngineProvider(system: ActorSystem) extends ConfigSSLEngineProvider(system) {
```
This approach won't scale. Extending `ConfigSSLEngineProvider` and then overriding `keyManagers` and `trustManagers` feels wrong.

Instead, I'm playing with the idea the an `SSLEngineProvider` is the composition of the following:

```
keyManagersProvider: (Config) => Array[KeyManager]
trustManagersProvider: (Config) => Array[TrustManager]
secureRandomProvider: (Config) => SecureRandom

sslContextFactory: (Array[KeyManager], Array[TrustManager], SecureRandom) => SSLContext
sslEngineFactory: (SSLContext, Config) => SSLEngine

// the following would be used in `verifyClientSession` and `verifyServerSession`, 
// e.g. to authorize the remote 
sslSessionVerifiers: Array[SSLSessionVerifiers]
```

With the above in mind, instead of building custom use cases via inheritance as I did here, users coudl compose more easily.

ignasi35 replied:
Note to self: add a `reference.conf`

patriknw replied:
it will be very rare to implement custom `SSLEngineProvider` so we shouldn't over-engineer with too much api and flexibility

patriknw replied:
extracting common code to shared internals might be doable, especially if you move this alongside the other in akka-remote.

patriknw replied:
that said, I'm not against a few more extension points but it must not turn into a new SslConfig. We deliberately kept things rather raw for SSLEngineProvider because more important to use well known standard Java api for these things even though it's ux is terrible.

ignasi35 replied:
> it must not turn into a new SslConfig

I agree with that 120%

patriknw commented on project/Dependencies.scala:
```diff
@@ -83,6 +83,8 @@ object Dependencies {
     // Added explicitly for when artery tcp is used
     val agrona = "org.agrona" % "agrona" % agronaVersion // ApacheV2
 
+    val asnOne = "com.hierynomus" % "asn-one" % "0.4.0"
```
add the `// ApacheV2` comment

(https://gitlab.com/hierynomus/asn-one/-/blob/master/LICENSE)

patriknw commented on akka-pem-reader/src/main/scala/akka/pki/pem/PEMDecoder.scala:
```diff
@@ -0,0 +1,60 @@
+    }
+  }
+
+  case class PEMData(label: String, bytes: Array[Byte])
```
Can we use `ByteString` or will it always be used as a byte array anyway?

ignasi35 replied:
Currently, the only usage of that field expects an `Array[Byte]` so I think it's better as it is.

patriknw commented on akka-pem-reader/src/main/scala/akka/pki/pem/PEMDecoder.scala:
```diff
@@ -0,0 +1,60 @@
+            Left(s"Error decoding base64 data from PEM data: ${iae.getMessage}")
+        }
+
+      case _ => Left("Not a PEM encoded data")
```
throw exceptions instead? then we have javadsl of this for free

patriknw commented on akka-pem-reader/src/main/scala/akka/pki/pem/PEMPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,107 @@
+object PEMPrivateKeyLoader {
+
+  def load(keyFile: String): Either[String, PrivateKey] = {
+    val pemData = new String(Files.readAllBytes(new File(keyFile).toPath))
```
that is using platform's default charset, which is bad idea in general

patriknw commented on akka-pem-reader/src/main/scala/akka/pki/pem/PEMPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,107 @@
+
+object PEMPrivateKeyLoader {
+
+  def load(keyFile: String): Either[String, PrivateKey] = {
```
should have docs about that it's using blocking file IO

ignasi35 committed:
Adds akka-pki

ignasi35 committed:
Adds PEM Loader

Co-Authored-By: James Roper <james@jazzy.id.au>

ignasi35 committed:
PR comments and other cleanups

ignasi35 commented:
I've rewritten this with several things in mind:

1. the module is named `akka-pki`
2. split commits to keep attribution (hat tip @jroper)
3. api docs
4. introduced a custom exception
5. reviewed naming and [used `PEM` when it's PEM and `DER` when it' DER](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail)
6. opted out from using `case class` for `DERData` 
7. decoupled `PEMDecoder` from `DERPrivateKeyLoader`

I've removed the sample implementation of a sub-class of `ConfigSSLEngineProvider` and removed the dependency between `akka-pki` and `akka-remote`.

I'd rather merge this as in and implement an example of an `SSLEngineProvider` using PEM inputs in a separate PR.

patriknw commented on akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,122 @@
+import com.hierynomus.asn1.types.constructed.ASN1Sequence
+import com.hierynomus.asn1.types.primitive.ASN1Integer
+
+class PEMLoadingException(message: String, cause: Throwable) extends RuntimeException(message, cause) {
```
`final class`

patriknw commented on akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  private def loadPkcs1PrivateKey(bytes: Array[Byte]) = {
+    val derInputStream = new ASN1InputStream(new DERDecoder, bytes)
```
no need to `close` it?

ignasi35 replied:
We're reading data from memory (which internally will use a `ByteArrayInputStream`) but closing won't harm.

patriknw commented on akka-pki/src/main/scala/akka/pki/pem/PEMDecoder.scala:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  @ApiMayChange
+  class DERData(val label: String, val bytes: Array[Byte])
```
`final class`

patriknw commented on akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,122 @@
+   * Converts the DER payload in [[PEMDecoder.DERData]] into a [[java.security.PrivateKey]]. The received DER
+   * data must be a valid PKCS#1 (identified in PEM as "RSA PRIVATE KEY") or non-ecrypted PKCS#8 (identified
+   * in PEM as "PRIVATE KEY").
+   */
```
document that it may throw PEMLoadingException

patriknw commented on build.sbt:
```diff
@@ -314,6 +315,14 @@ lazy val protobufV3 = akkaModule("akka-protobuf-v3")
+lazy val pki =
+  akkaModule("akka-pki")
+    .dependsOn(actor) // this dependency only exists for "@ApiMayChange"
+    .settings(Dependencies.pki)
```
missing AutomaticModuleName

(let's ignore osgi until someone complains)

patriknw reviewed (approved):
LGTM, only a few nitpicks...

ignasi35 committed:
PR comments (also, Fix unidoc issue)

ignasi35 commented:
> Test FAILed.

The failure seems unrelated to the changes in the PR:

```
...
[05-12 20:59:35.989] [error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala:201:29: value getEventQueue is not a member of org.slf4j.helpers.SubstituteLoggerFactory
[05-12 20:59:35.989] [error]     substituteLoggerFactory.getEventQueue
[05-12 20:59:35.989] [error]                             ^
[05-12 20:59:35.992] [error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala:218:29: value getEventQueue is not a member of org.slf4j.helpers.SubstituteLoggerFactory
[05-12 20:59:35.992] [error]     substituteLoggerFactory.getEventQueue.clear()
[05-12 20:59:35.992] [error]                             ^
[05-12 21:00:11.318] [info] No documentation generated with unsuccessful compiler run
[05-12 21:00:11.319] [error] two errors found
...

[05-12 21:24:04.072] [error] (Scalaunidoc / doc) Scaladoc generation failed
[error] Total time: 2821 s (47:01), completed May 12, 2020 11:24:04 PM
```

I can reproduce the error locally invoking `unidoc`, but when I invoke `ValidatePR / unidoc` I get a similar error in a different class. I'm not sure what's going on here...

ignasi35 commented:
Hmmm, I just found #18401 but I wonder how a project as simple as `akka-pki` which noone depends on may cause a failure running `doc` over `akka/actor/testkit/typed/internal/StubbedActorContext`.

jrudolph commented on build.sbt:
```diff
@@ -314,6 +316,15 @@ lazy val protobufV3 = akkaModule("akka-protobuf-v3")
+    .settings(AutomaticModuleName.settings("akka.pki"))
+    .settings(
+      // The akka-pki artifact was added in Akka 2.6.2, no MiMa checks yet.
+      mimaPreviousArtifacts := Set.empty)
```
Shouldn't we set it up correctly right away in Mima.scala? If we want to ignore binary compatibility for now, let's use `disablePlugin(MimaPlugin)` instead for clarity.

jrudolph commented:
> Hmmm, I just found #18401 but I wonder how a project as simple as `akka-pki` which noone depends on may cause a failure running `doc` over `akka/actor/testkit/typed/internal/StubbedActorContext`.

One guess would be, that unidoc aggregates code from all projects and tries to build documentation in a single run and somehow mangles transitive dependencies uncleanly.

Indeed, this one shows:

```
akka > show scalaunidoc:unidoc::fullClasspath
[warn] Pemfile /home/johannes/.ssh/id_rsa_gustav.pem doesn't exist.
[warn] Pemfile /home/johannes/.ssh/id_rsa_gustav.pem doesn't exist.
[info] Strategy 'discard' was applied to 2 files (Run the task at debug level to see details)
[info] Assembly up to date: /home/johannes/git/opensource/akka-streams/akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-actor-typed/target/scala-2.12/CompileJdk9-classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-remote/target/scala-2.12/CompileJdk9-classes)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/io/aeron/aeron-client/1.27.0/aeron-client-1.27.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/io/aeron/aeron-driver/1.27.0/aeron-driver-1.27.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/agrona/agrona/1.4.1/agrona-1.4.1.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-protobuf-v3/target/scala-2.12/stripped/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-sharding-typed/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-typed/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-remote/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-stream/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-tools/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-distributed-data/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-sharding/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-persistence/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-multi-node-testkit/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-metrics/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/../akka-protobuf-v3/target/scala-2.12/akka-protobuf-v3-assembly-2.6.5+45-c298b657.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/ow2/asm/asm/5.0.3/asm-5.0.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/ow2/asm/asm-analysis/5.0.3/asm-analysis-5.0.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/ow2/asm/asm-commons/5.0.3/asm-commons-5.0.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/ow2/asm/asm-tree/5.0.3/asm-tree-5.0.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/ow2/asm/asm-util/5.0.3/asm-util-5.0.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/hierynomus/asn-one/0.4.0/asn-one-0.4.0.jar)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-actor-typed/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-actor/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-slf4j/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-typed/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-remote/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-stream/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-tools/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-coordination/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-distributed-data/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-cluster-sharding/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-persistence/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-testkit/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-persistence-typed/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/git/opensource/akka-streams/akka-actor-testkit-typed/target/scala-2.12/classes)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/typesafe/config/1.4.0/config-1.4.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/google/guava/guava/21.0/guava-21.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-annotations/2.10.4/jackson-annotations-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.10.4/jackson-core-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-databind/2.10.4/jackson-databind-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/dataformat/jackson-dataformat-cbor/2.10.4/jackson-dataformat-cbor-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/datatype/jackson-datatype-jdk8/2.10.4/jackson-datatype-jdk8-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/datatype/jackson-datatype-jsr310/2.10.4/jackson-datatype-jsr310-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/module/jackson-module-parameter-names/2.10.4/jackson-module-parameter-names-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/module/jackson-module-paranamer/2.10.4/jackson-module-paranamer-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/module/jackson-module-scala_2.12/2.10.4/jackson-module-scala_2.12-2.10.4.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/sun/activation/javax.activation/1.2.0/javax.activation-1.2.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/jnr/jffi/1.2.18/jffi-1.2.18-native.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/jnr/jffi/1.2.18/jffi-1.2.18.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/jnr/jnr-a64asm/1.0.0/jnr-a64asm-1.0.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/jnr/jnr-constants/0.9.12/jnr-constants-0.9.12.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/jnr/jnr-ffi/2.1.9/jnr-ffi-2.1.9.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/jnr/jnr-x86asm/1.0.2/jnr-x86asm-1.0.2.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/junit/junit/4.13/junit-4.13.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/iq80/leveldb/leveldb/0.12/leveldb-0.12.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/iq80/leveldb/leveldb-api/0.12/leveldb-api-0.12.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/fusesource/leveldbjni/leveldbjni-all/1.8/leveldbjni-all-1.8.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/lmdbjava/lmdbjava/0.7.0/lmdbjava-0.7.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/ch/qos/logback/logback-classic/1.2.3/logback-classic-1.2.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/ch/qos/logback/logback-core/1.2.3/logback-core-1.2.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/io/netty/netty/3.10.6.Final/netty-3.10.6.Final.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/osgi/org.osgi.compendium/5.0.0/org.osgi.compendium-5.0.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/osgi/org.osgi.core/6.0.0/org.osgi.core-6.0.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/thoughtworks/paranamer/paranamer/2.8/paranamer-2.8.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/reactivestreams/reactive-streams/1.0.3/reactive-streams-1.0.3.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-java8-compat_2.12/0.8.0/scala-java8-compat_2.12-0.8.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.12.11/scala-library-2.12.11.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-parser-combinators_2.12/1.1.2/scala-parser-combinators_2.12-1.1.2.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.11/scala-reflect-2.12.11.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.2.0/scala-xml_2.12-1.2.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalactic/scalactic_2.12/3.1.1/scalactic_2.12-3.1.1.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalatest/scalatest_2.12/3.1.1/scalatest_2.12-3.1.1.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/io/kamon/sigar-loader/1.6.6-rev002/sigar-loader-1.6.6-rev002.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/ghik/silencer-lib_2.12.11/1.6.0/silencer-lib_2.12.11-1.6.0.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/slf4j/slf4j-api/1.7.13/slf4j-api-1.7.13.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/org/slf4j/slf4j-api/1.7.30/slf4j-api-1.7.30.jar)
[info] * Attributed(/home/johannes/.coursier/cache/v1/https/repo1.maven.org/maven2/com/typesafe/ssl-config-core_2.12/0.4.1/ssl-config-core_2.12-0.4.1.jar)
[success] Total time: 3 s, completed May 14, 2020 10:35:03 AM
```

So, akka-pki pulls in asnOne which depends on slf4j-api 1.7.13. Unidoc doesn't seem to do conflict resolution on the transitive dependencies, so both slf4j-api 1.7.13 and 1.7.30 end up on the classpath. I guess a short-term solution is to add an explicit dependency on the later version of slf4j to the akka-pki project.

jrudolph commented on project/Dependencies.scala:
```diff
@@ -195,6 +197,8 @@ object Dependencies {
 
   val actorTestkitTyped = l ++= Seq(Provided.logback, Provided.junit, Provided.scalatest, Test.scalatestJUnit)
 
+  val pki = l ++= Seq(asnOne, Test.scalatest)
```
This fixes the unidoc issue for me:

```suggestion
  val pki = l ++=
      Seq(
        asnOne,
        // pull up slf4j version from the one provided transitively in asnOne to fix unidoc
        Compile.slf4jApi % "provided",
        Test.scalatest)
```

ignasi35 replied:
thanks @jrudolph

seglo commented on akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,130 @@
+
+  /**
+   * Converts the DER payload in [[PEMDecoder.DERData]] into a [[java.security.PrivateKey]]. The received DER
+   * data must be a valid PKCS#1 (identified in PEM as "RSA PRIVATE KEY") or non-ecrypted PKCS#8 (identified
```
```suggestion
   * data must be a valid PKCS#1 (identified in PEM as "RSA PRIVATE KEY") or non-encrypted PKCS#8 (identified
```

seglo commented on akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,130 @@
+    // Here's the specification: https://tools.ietf.org/html/rfc3447#appendix-A.1.2
+    val sequence = {
+      try {
+        derInputStream.readObject[ASN1Sequence]()
```
Would it make sense to nest possible exceptions in a new `PEMLoadingException`?

ignasi35 replied:
I thought about it a bit but the `PEMLoadingException` will only be thrown if an unsupported content (e.g. `CERTIFICATE` or `ENCRYPTED PRIVATE KEY`) is loaded. Also when converting from PEM to DER. But once we have the `Array[Byte]` with DER only other exceptions (e.g.`IllegalArgument` or anything thrown by underlying libraries) is thrown.

seglo commented on akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,130 @@
+        otherPrimeInfos)
+    }
+
+    val keyFactory = KeyFactory.getInstance("RSA")
```
I assume it's safe to always generate RSA keys.

ignasi35 replied:
This is reading PKCS#1 si it must be RSA.

ignasi35 replied:
I am paranoid so I checked [the PKCS#8 case too](https://github.com/openjdk/jdk/blob/6bab0f539fba8fb441697846347597b4a0ade428/src/java.base/share/classes/sun/security/rsa/RSAKeyFactory.java#L38-L66).

seglo commented on akka-pki/src/main/scala/akka/pki/pem/PEMDecoder.scala:
```diff
@@ -0,0 +1,76 @@
+    val laxtextualmessage = raw"""\s*$preeb($laxbase64text)$posteb\s*"""
+
+    laxtextualmessage.r
+  }
```
whoa

ignasi35 replied:
I know.

jroper replied:
I *love* that this regex has made it into Akka. I can truly say I've had a bad influence on Akka now.

jroper replied:
And what's even better is that git blame says that Ignasi wrote it, so it'll never come back to me when someone has to fix a bug in it :grin:

jroper replied:
Oh wait, I just saw, @ignasi35 managed to import the commit from the other repo but keep the author as me... Oh well. I wrote this code. Deal with it.

seglo reviewed (approved):
I'm new to most of this, but LGTM. Found a typo, made some exclamations!!

jroper commented on akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala:
```diff
@@ -0,0 +1,130 @@
+        coefficient)
+    } else {
+      // Does anyone even use multi-primes? Who knows, maybe this code will never be used. Anyway, I guess it will work,
+      // the spec isn't exactly complicated.
```
For posterity, I tried to write tests for this (there's a comment below), but as far as I can see, while the JDK does have the `RSAMultiPrimePrivateCrtKeySpec` to represent multiprime private keys, as far as I could see, the built in JSSE provider ignores the additional primes, so there's no way to test that using this parsed key spec works, and the built in PKCS8 parser throws an exception if you pass it a multi prime key, so I couldn't compare the output of this to a PKCS8 parsed key. The tests just test that multi prime keys are parsed without throwing an exception, whether what's parsed is correct is untested because there was no real way to test it. I suppose if you use a different JSSE provider, you might get multi prime support, and then this code will be useful if you use multiprime keys, but the fact that the JDK doesn't implement it suggests to me that it's a feature that's never used, so this code is possibly effectively dead code.
